twake/backend/node/src/services/previews/services/links/engine/index.ts

```diff
@@ -0,0 +1,20 @@
+import { Initializable } from "../../../../../core/platform/framework";
+import { LinkPreviewProcessor } from "./service";
+import gr from "../../../../global-resolver";
+import { LinkPreviewFinishedProcessor } from "./finished";
+import { Message } from "../../../../../services/messages/entities/messages";
+import Repository from "../../../../../core/platform/services/database/services/orm/repository/repository";
+
+export class LinkPreviewEngine implements Initializable {
+  private repository: Repository<Message>;
+  async init(): Promise<this> {
+    this.repository = await gr.database.getRepository<Message>("messages", Message);
```

Comment:
This and also the whole LinkPreviewFinishedProcessor should be in the message service.
The idea is to have this communication through rabbitmq: `messageService <-> previewService` So previewService should not know anything about messages, just receive a task and return the result asynchrounously.

Reply:
Doing so, back on the messages service you can also call the websocket API in order to make the frontend update the message automatically in realtime like done in this example from `twake/backend/node/src/services/messages/services/engine/processors/channel-view/index.ts` and `twake/backend/node/src/services/messages/services/engine/processors/user-inbox/index.ts`, **it could be good to regroup them in a single helper function by the way**:

```
//Publish message in realtime
      const room = `/companies/${participant.company_id}/workspaces/${participant.workspace_id}/channels/${participant.id}/feed`;
      const type = "message";
      const entity = message.resource;
      const context = message.context;
      localEventBus.publish("realtime:publish", {
        topic: message.created
          ? RealtimeEntityActionType.Created
          : RealtimeEntityActionType.Updated,
        event: {
          type: type,
          room: ResourcePath.get(room),
          resourcePath: getThreadMessagePath(context as ThreadExecutionContext) + "/" + entity.id,
          entity: entity,
          result: message.created
            ? new CreateResult<Message>(type, entity)
            : new UpdateResult<Message>(type, entity),
        },
      } as RealtimeLocalBusEvent<Message>);

```